# Task 2: Data Discovery, Profiling and Cleaning

This notebook profiles the extracted PNU research datasets, identifies data quality issues, and applies cleaning steps to prepare the data for further processing.

## 1. Import Required Libraries

This section imports the Python libraries required for loading, analyzing, and cleaning the extracted datasets.

In [2]:
# Import pandas for data manipulation and analysis
import pandas as pd

# Import os for handling file paths
import os

## 2. Load Extracted Data

This section loads the extracted PNU datasets generated during the extraction stage.
The files are loaded from the interim folder without modifying the original raw data.

In [3]:
# Define the path where extracted datasets are stored
interim_path = "../data/interim/"

# Load extracted PNU datasets
pnu_2023 = pd.read_csv(interim_path + "PNU_2023_extracted.csv")
pnu_2024 = pd.read_csv(interim_path + "PNU_2024_extracted.csv")
pnu_2025 = pd.read_csv(interim_path + "PNU_2025_extracted.csv")

# Display the number of rows loaded from each dataset
print("PNU 2023:", pnu_2023.shape)
print("PNU 2024:", pnu_2024.shape)
print("PNU 2025:", pnu_2025.shape)

PNU 2023: (4589, 7)
PNU 2024: (4225, 7)
PNU 2025: (1664, 7)


## 3. Combine Extracted Datasets

The extracted datasets from different years are combined into a single dataframe to perform consistent profiling and cleaning steps.

In [4]:
# Combine PNU datasets from different years
pnu_data = pd.concat(
    [pnu_2023, pnu_2024, pnu_2025],
    ignore_index=True
)

# Display the combined dataset size
pnu_data.shape

(10478, 7)

## 4. Data Profiling

This section examines the structure and quality of the combined dataset by checking column names, data types, missing values, unique values, and sample records.

In [5]:
# Display column names of the combined dataset
pnu_data.columns

Index(['Publication Year', 'Month', 'Category', 'college', 'Research field ',
       'Title', 'Source Title'],
      dtype='str')

### 4.1 Profiling Summary

This section generates a summary of each column including data type, missing values, unique values, and sample values.

In [6]:
# Create a profiling summary table for the dataset

profiling_summary = pd.DataFrame({
    "column": pnu_data.columns,
    "dtype": pnu_data.dtypes.values,
    "null_count": pnu_data.isnull().sum().values,
    "percent_null": (pnu_data.isnull().sum().values / len(pnu_data) * 100).round(2),
    "unique_count": pnu_data.nunique().values,
    "sample_value": [pnu_data[col].dropna().iloc[0] for col in pnu_data.columns]
})

# Display profiling summary
profiling_summary

,column,dtype,null_count,percent_null,unique_count,sample_value
0,Publication Year,int64,0,0.0,3,2023
1,Month,str,0,0.0,12,March
2,Category,str,0,0.0,6,scopus
3,college,str,0,0.0,46,Arabic Language Teaching Institute for Non-Ara...
4,Research field,str,0,0.0,4,Humanitarian
5,Title,str,0,0.0,9404,Sexual workplace violence in the health sector...
6,Source Title,str,0,0.0,2551,Computer Systems Science and Engineering


### 4.2 Sample Records

This section displays sample records to understand the structure and content of the dataset.

In [7]:
# Display sample rows from the dataset
pnu_data.head()

,Publication Year,Month,Category,college,Research field,Title,Source Title
0,2023,March,scopus,Arabic Language Teaching Institute for Non-Ara...,Humanitarian,Sexual workplace violence in the health sector...,Computer Systems Science and Engineering
1,2023,April,SCOPUS,Arabic Language Teaching Institute for Non-Ara...,Humanitarian,Spatial Analysis of Accessibility to Urban Gre...,Intelligent Automation and Soft Computing
2,2023,April,SCOPUS,Arabic Language Teaching Institute for Non-Ara...,Humanitarian,"Association of CCND1 (c.723G > A, rs9344) vari...",Intelligent Automation and Soft Computing
3,2023,August,scopus,Arabic Language Teaching Institute for Non-Ara...,Humanitarian,Neuroprotective Effect of Barbaloin on Strepto...,Computer Systems Science and Engineering
4,2023,December,Wos,Arabic Language Teaching Institute for Non-Ara...,Humanitarian,Adoption of antenatal care conversation mappin...,PEERJ COMPUTER SCIENCE


### 4.3 Duplicate Records Check

This section checks for duplicated records in the combined dataset.

In [8]:
# Check the number of duplicate rows in the dataset
duplicate_count = pnu_data.duplicated().sum()

duplicate_count

np.int64(5)

## 4.4 Data Quality Issues Identified

The profiling step identified the following data quality issues:

### 1. Duplicate Records
- Duplicate records were identified and removed during the cleaning stage.
- Duplicate records were removed during the cleaning stage.

### 2. Text Standardization
- Some text values had inconsistent capitalization (for example, WOS and wos).
- Text fields were standardized by applying lowercase conversion and trimming spaces.

### 3. Column Name Formatting
- Column names contained spaces and inconsistent formatting.
- Column names were renamed using snake_case format.

### 4.5 Data Type Validation

This section checks the data types of each column to identify any inconsistencies.

In [9]:
# Display data types of all columns
pnu_data.dtypes

Publication Year    int64
Month                 str
Category              str
college               str
Research field        str
Title                 str
Source Title          str
dtype: object

## 5. Data Cleaning

This section applies cleaning steps to improve data quality by removing duplicates, standardizing text values, and preparing the dataset for further processing.

### 5.1 Remove Duplicate Records

Duplicate records identified during profiling are removed to avoid repeated entries in the dataset.

In [10]:
# Remove duplicate records
pnu_cleaned = pnu_data.drop_duplicates()

# Display the cleaned dataset size
pnu_cleaned.shape

(10473, 7)

### 5.2 Standardize Text Values

This step standardizes text columns by removing extra spaces and converting text values to lowercase to ensure consistency.

In [11]:
# Standardize text columns
text_columns = pnu_cleaned.select_dtypes(include="object").columns

for col in text_columns:
    pnu_cleaned[col] = (
        pnu_cleaned[col]
        .str.strip()
        .str.lower()
    )

# Display sample after text standardization
pnu_cleaned.head()

C:\Users\arrsa\AppData\Local\Temp\ipykernel_6980\364841719.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  text_columns = pnu_cleaned.select_dtypes(include="object").columns


,Publication Year,Month,Category,college,Research field,Title,Source Title
0,2023,march,scopus,arabic language teaching institute for non-ara...,humanitarian,sexual workplace violence in the health sector...,computer systems science and engineering
1,2023,april,scopus,arabic language teaching institute for non-ara...,humanitarian,spatial analysis of accessibility to urban gre...,intelligent automation and soft computing
2,2023,april,scopus,arabic language teaching institute for non-ara...,humanitarian,"association of ccnd1 (c.723g > a, rs9344) vari...",intelligent automation and soft computing
3,2023,august,scopus,arabic language teaching institute for non-ara...,humanitarian,neuroprotective effect of barbaloin on strepto...,computer systems science and engineering
4,2023,december,wos,arabic language teaching institute for non-ara...,humanitarian,adoption of antenatal care conversation mappin...,peerj computer science


### 5.3 Rename Columns to snake_case

This step renames columns using snake_case format to improve consistency and compatibility with data processing tools.

In [37]:
# Rename columns to snake_case format

pnu_cleaned.columns = (
    pnu_cleaned.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_", regex=False)
)

# Display updated column names
pnu_cleaned.columns

Index(['research_id', 'university', 'title', 'authors', 'publication_year',
       'publication_date', 'abstract', 'research_field', 'tech_category',
       'journal', 'doi', 'url', 'source'],
      dtype='str')

### 5.4 Handle Missing Values

This step checks for missing values after cleaning and applies appropriate handling if required.

In [13]:
# Check missing values after cleaning
missing_values = pnu_cleaned.isnull().sum()

# Display columns with missing values
missing_values

publication_year    0
month               0
category            0
college             0
research_field      0
title               0
source_title        0
dtype: int64

### 5.5 Validate Cleaned Dataset

This section validates the cleaned dataset by checking its structure and confirming that cleaning steps were applied successfully.

In [14]:
# Display final dataset information after cleaning
pnu_cleaned.info()

<class 'pandas.DataFrame'>
Index: 10473 entries, 0 to 10477
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   publication_year  10473 non-null  int64
 1   month             10473 non-null  str  
 2   category          10473 non-null  str  
 3   college           10473 non-null  str  
 4   research_field    10473 non-null  str  
 5   title             10473 non-null  str  
 6   source_title      10473 non-null  str  
dtypes: int64(1), str(6)
memory usage: 654.6 KB


## Common Schema Mapping

This section maps the cleaned PNU dataset into the project common schema required for integration with other university datasets.

In [15]:
# Create common schema columns required for integration

pnu_cleaned["research_id"] = (
    "PNU-" + pnu_cleaned.index.astype(str)
)

pnu_cleaned["university"] = "PNU"

pnu_cleaned["authors"] = pd.NA

pnu_cleaned["publication_date"] = pd.NaT

pnu_cleaned["abstract"] = pd.NA

pnu_cleaned["tech_category"] = pd.NA

pnu_cleaned["journal"] = pnu_cleaned["source_title"]

pnu_cleaned["doi"] = pd.NA

pnu_cleaned["url"] = pd.NA

pnu_cleaned["source"] = "PNU Open Data"


# Reorder columns according to common schema

pnu_cleaned = pnu_cleaned[
    [
        "research_id",
        "university",
        "title",
        "authors",
        "publication_year",
        "publication_date",
        "abstract",
        "research_field",
        "tech_category",
        "journal",
        "doi",
        "url",
        "source"
    ]
].copy()


print("Final rows:", len(pnu_cleaned))
print("Final columns:", len(pnu_cleaned.columns))

pnu_cleaned.head()

Final rows: 10473
Final columns: 13


,research_id,university,title,authors,publication_year,publication_date,abstract,research_field,tech_category,journal,doi,url,source
0,PNU-0,PNU,sexual workplace violence in the health sector...,<NA>,2023,NaT,<NA>,humanitarian,<NA>,computer systems science and engineering,<NA>,<NA>,PNU Open Data
1,PNU-1,PNU,spatial analysis of accessibility to urban gre...,<NA>,2023,NaT,<NA>,humanitarian,<NA>,intelligent automation and soft computing,<NA>,<NA>,PNU Open Data
2,PNU-2,PNU,"association of ccnd1 (c.723g > a, rs9344) vari...",<NA>,2023,NaT,<NA>,humanitarian,<NA>,intelligent automation and soft computing,<NA>,<NA>,PNU Open Data
3,PNU-3,PNU,neuroprotective effect of barbaloin on strepto...,<NA>,2023,NaT,<NA>,humanitarian,<NA>,computer systems science and engineering,<NA>,<NA>,PNU Open Data
4,PNU-4,PNU,adoption of antenatal care conversation mappin...,<NA>,2023,NaT,<NA>,humanitarian,<NA>,peerj computer science,<NA>,<NA>,PNU Open Data


## Technical Research Filtering

This section filters PNU research records related to technical fields
before metadata enrichment.
The filtering is based on research field, college, and title keywords.

In [18]:
# Create a copy for technical research filtering

pnu_technical = pnu_cleaned.copy()


# Define technical keywords

technical_keywords = [
    "computer",
    "computing",
    "artificial intelligence",
    "ai",
    "machine learning",
    "data",
    "software",
    "engineering",
    "technology",
    "information",
    "cyber",
    "security",
    "network",
    "system",
    "digital",
    "automation",
    "robot",
    "smart"
]


# Combine text columns for searching

search_columns = [
    "title",
    "research_field",
    "journal"
]


pnu_technical["search_text"] = (
    pnu_technical[search_columns]
    .fillna("")
    .astype(str)
    .agg(" ".join, axis=1)
    .str.lower()
)


# Filter technical research

pattern = "|".join(technical_keywords)

pnu_technical = pnu_technical[
    pnu_technical["search_text"].str.contains(
        pattern,
        na=False
    )
].copy()


# Remove helper column

pnu_technical.drop(
    columns="search_text",
    inplace=True
)


print("Original PNU records:", len(pnu_cleaned))
print("Technical research records:", len(pnu_technical))

Original PNU records: 10473
Technical research records: 4193


## Crossref Metadata Enrichment

This section retrieves missing publication metadata from Crossref API,
including authors, DOI, URL, publication date, abstract, and journal
information for technical research records.

In [29]:
# Import required libraries

import requests
import time
from tqdm import tqdm


def get_crossref_data(title):

    api_url = "https://api.crossref.org/works"

    params = {
        "query.title": title,
        "rows": 1
    }

    try:
        response = requests.get(
            api_url,
            params=params,
            timeout=10
        )

        if response.status_code != 200:
            return pd.Series({
                "authors": pd.NA,
                "doi": pd.NA,
                "url": pd.NA,
                "publication_date": pd.NaT,
                "abstract": pd.NA,
                "journal": pd.NA
            })

        items = response.json()["message"]["items"]

        if len(items) == 0:
            return pd.Series({
                "authors": pd.NA,
                "doi": pd.NA,
                "url": pd.NA,
                "publication_date": pd.NaT,
                "abstract": pd.NA,
                "journal": pd.NA
            })


        work = items[0]


        # Extract authors
        authors = "; ".join(
            [
                f"{a.get('given','')} {a.get('family','')}".strip()
                for a in work.get("author", [])
            ]
        )


        # Extract date
        date_parts = (
            work.get("published-print", {})
            .get("date-parts")
            or work.get("published-online", {})
            .get("date-parts")
        )

        publication_date = pd.NaT

        if date_parts:
            parts = date_parts[0]

            if len(parts) >= 3:
                publication_date = pd.Timestamp(
                    year=parts[0],
                    month=parts[1],
                    day=parts[2]
                )


        return pd.Series({
            "authors": authors if authors else pd.NA,
            "doi": work.get("DOI"),
            "url": work.get("URL"),
            "publication_date": publication_date,
            "abstract": work.get("abstract"),
            "journal": (
                work.get("container-title", [pd.NA])[0]
                if work.get("container-title")
                else pd.NA
            )
        })


    except Exception:
        return pd.Series({
            "authors": pd.NA,
            "doi": pd.NA,
            "url": pd.NA,
            "publication_date": pd.NaT,
            "abstract": pd.NA,
            "journal": pd.NA
        })

In [30]:
# Create enrichment dataset

pnu_enriched = pnu_technical.copy()


crossref_results = []

for title in tqdm(pnu_enriched["title"]):

    crossref_results.append(
        get_crossref_data(title)
    )

    time.sleep(0.2)


crossref_df = pd.DataFrame(crossref_results)


# Update metadata columns

pnu_enriched[
    [
        "authors",
        "doi",
        "url",
        "publication_date",
        "abstract",
        "journal"
    ]
] = crossref_df


print("Crossref enrichment completed")
print("Records:", len(pnu_enriched))

100%|██████████| 4193/4193 [1:47:11<00:00,  1.53s/it]   

Crossref enrichment completed
Records: 4193


In [31]:
pnu_enriched[
    ["authors","doi","url"]
].notna().sum()

authors    1544
doi        1552
url        1552
dtype: int64

In [32]:
# Check mandatory fields availability after Crossref enrichment

mandatory_check = pnu_enriched[
    [
        "research_id",
        "title",
        "authors",
        "doi",
        "url"
    ]
].isna().sum()

mandatory_check

research_id       0
title             0
authors        2649
doi            2641
url            2641
dtype: int64

In [33]:
# Count records that satisfy mandatory requirements

valid_mandatory = pnu_enriched[
    pnu_enriched["authors"].notna()
    & pnu_enriched["doi"].notna()
    & pnu_enriched["url"].notna()
]

print("Records passing mandatory fields:", len(valid_mandatory))
print("Records rejected:", len(pnu_enriched) - len(valid_mandatory))

Records passing mandatory fields: 1544
Records rejected: 2649


In [34]:
# Check records that satisfy all mandatory fields

valid_records = pnu_enriched[
    pnu_enriched["research_id"].notna()
    & pnu_enriched["title"].notna()
    & pnu_enriched["publication_year"].notna()
    & pnu_enriched["doi"].notna()
    & pnu_enriched["url"].notna()
    & pnu_enriched["authors"].notna()
]

print("Valid records:", len(valid_records))
print("Rejected records:", len(pnu_enriched) - len(valid_records))

Valid records: 1544
Rejected records: 2649


In [35]:
# Separate validated and rejected records based on mandatory fields

pnu_validated = pnu_enriched[
    pnu_enriched["authors"].notna()
    & pnu_enriched["doi"].notna()
    & pnu_enriched["url"].notna()
].copy()


pnu_rejected = pnu_enriched[
    pnu_enriched["authors"].isna()
    | pnu_enriched["doi"].isna()
    | pnu_enriched["url"].isna()
].copy()


print("PNU total:", len(pnu_enriched))
print("PNU validated:", len(pnu_validated))
print("PNU rejected:", len(pnu_rejected))

PNU total: 4193
PNU validated: 1544
PNU rejected: 2649


In [36]:
# Add validation error reasons

def get_validation_error(row):

    errors = []

    if pd.isna(row["authors"]):
        errors.append("authors is missing")

    if pd.isna(row["doi"]):
        errors.append("doi is missing")

    if pd.isna(row["url"]):
        errors.append("url is missing")

    return "; ".join(errors)


pnu_rejected["validation_error"] = (
    pnu_rejected.apply(
        get_validation_error,
        axis=1
    )
)


pnu_rejected["validation_error"].value_counts()

validation_error
authors is missing; doi is missing; url is missing    2641
authors is missing                                       8
Name: count, dtype: int64

## Crossref Enrichment Results

Crossref API was used to retrieve missing metadata fields for technical PNU research records.

After enrichment:
- Total technical records processed: 4193
- Records with complete mandatory metadata: 1544
- Records missing mandatory metadata: 2649

Records missing mandatory fields were separated according to the project validation rules.

In [38]:
# Save validated and rejected datasets

pnu_validated.to_csv(
    "../data/interim/PNU_validated.csv",
    index=False
)


pnu_rejected.to_csv(
    "../data/interim/PNU_rejected.csv",
    index=False
)


print("Validation files saved successfully.")

Validation files saved successfully.


The enriched dataset was validated and separated into:

- data/interim/PNU_validated.csv
- data/interim/PNU_rejected.csv

The validated dataset follows the common schema with 13 columns.

## Task 2 Summary

The PNU dataset was profiled, cleaned, and prepared for the next pipeline stage.

Data cleaning steps included:
- Standardizing column formats.
- Removing unnecessary spaces from text fields.
- Handling missing values while preserving optional fields.
- Validating publication years and data types.
- Removing duplicate records where applicable.

The cleaned dataset was then mapped to the project common schema required for integration with other university datasets.

The enriched dataset was validated and separated into:

- `data/interim/PNU_validated.csv`
- `data/interim/PNU_rejected.csv`

The validated dataset follows the common schema with 13 columns.

The output follows the common schema with 13 columns:
- research_id
- university
- title
- authors
- publication_year
- publication_date
- abstract
- research_field
- tech_category
- journal
- doi
- url
- source